### ISSUE TITLE
Build failure in version 1.8.0

### ISSUE THREAD

**Issue by erikd** [serokell/log-warper#89]
```
src/System/Wlog/LoggerName.hs:40:15: error:
    • No instance for (Formatting.Buildable.Buildable LoggerName)
        arising from a use of ‘build’
    • In the expression: build
      In an equation for ‘loggerNameF’: loggerNameF = build
```

**Comment by chshersh** [comment]
@erikd Could you elaborate more on the way you build `log-warper`? I can build `log-warper-1.8.0` without problems. Which building tool are you using? What are dependencies versions?

**Comment by erikd** [comment]
I'm using the [mafia](https://github.com/haskell-mafia/mafia) build tool. Mafia is a thin wrapper around cabal.

I can see from your `.tavis.yml` file that you only test with Stack and from your cabal file that you only specify lower bounds. This leaves dependencies somewhat under constrained and it is therefore possible to make this un-buildable with cabal (and hence mafia).

Let me do some more investigation.

**Comment by erikd** [comment]
Its pretty trivial to reproduce this just using cabal:
```
git clone https://github.com/serokell/log-warper
cd log-warper
git checkout v1.8.0 -b branch-v1.8.0
cabal sandbox init
cabal install --dependencies-only --enable-tests
cabal build
```
which results in:
```
src/System/Wlog/LoggerName.hs:40:15: error:
    • No instance for (Formatting.Buildable.Buildable LoggerName)
        arising from a use of ‘build’
    • In the expression: build
      In an equation for ‘loggerNameF’: loggerNameF = build
```
For packages that are destined to be uploaded to Hackage, I would actually recommend against using Stack to build them in Travis, for exactly this reason.

I've you're interested, I could modify your `.travis.yml` file to use Cabal instead of Stack and then fix the version bounds.

**Comment by erikd** [comment]
Building the `v1.8.0` tag required the following patch to the cabal file:
```
diff --git a/log-warper.cabal b/log-warper.cabal
index 08ccf56..0da2144 100644
--- a/log-warper.cabal
+++ b/log-warper.cabal
@@ -51,7 +51,7 @@ library
                      , extra                >= 1.4.10
                      , filepath             >= 1.4.1
                      , fmt                  >= 0.5.0.0
-                     , formatting           >= 6.2.2
+                     , formatting           >= 6.2.2        && < 6.3
                      , hashable             >= 1.2.4.0
                      , lens                 >= 4.14
                      , monad-control        >= 1.0.1.0
@@ -64,7 +64,7 @@ library
                      , time                 >= 1.6.0.1
                      , transformers         >= 0.5.2
                      , transformers-base    >= 0.4.4
-                     , universum            >= 0.7
+                     , universum            >= 0.7          && < 1.0
                      , unordered-containers >= 0.2.7.1
                      , vector               >= 0.12
                      , yaml                 >= 0.8.20
```

**Comment by chshersh** [comment]
@erikd Thanks for finding the reason of the bug! But I'm not sure that version `1.8.0` requires a fix. Latest version of `log-warper` is `1.8.4`. It doesn't depend on `formatting` and uses `universum-1.0.2` package.

> For packages that are destined to be uploaded to Hackage

Well, every package is destined to be uploaded to Hackage :smirk: But `log-warper` is also curated on Stackage. All dependencies' bounds are fixed. I just need to upgrade package from time to time if something breaks on Stackage Nightly. Entire Serokell team uses `stack` as a build tool. So it's much easier to manage dependencies just by relying on stack and not specifying lower and upper bounds manually. I prefer option 3 from this SO answer:

* https://stackoverflow.com/a/47795973/2900502

That's why Travis CI uses `stack` — it allows to find potential problems with Stackage much sooner.

**Comment by erikd** [comment]
> But I'm not sure that version 1.8.0 requires a fix. Latest version of log-warper is 1.8.4

Well I ran into this problem building Cardano which does require 1.8.0.

What do you think of building in Travis with stack and cabal?

**Comment by chshersh** [comment]
> Well I ran into this problem building Cardano which does require 1.8.0.

At some moment dependencies in `cardano-sl` will be bumped up. Also, instead of changing dependencies' constraints in `log-warper` you can also specify package bounds in `cardano-sl` itself:

* https://github.com/input-output-hk/cardano-sl/blob/cf593ff80dd31917c9a868cd10aa6a719419a91c/lib/cardano-sl.cabal#L183

Moreover, `cardano-sl` currently uses at least two formatting libraries: `formatting` and `fmt` (which is sad). At some moment `formatting` should disappear (I hope). So, probably, there's a reason to not care much about `formatting`... But dependencies for `cardano-sl` should be optimized anyways.

> What do you think of building in Travis with stack and cabal?

Not sure about this... :thinking: Current Travis settings for `log-warper` (and other Serokell repositories) were developed by our DevOps team to achieve the best possible results for Haskell builds. Blog post article about this is coming soon. As I mentioned earlier, we don't really use `cabal` build tool in our team. I'm not sure it worth it to complicate building and project management processes. I can always upload documentation manually to Hackage so users can see haddocks even if Hackage build fails.

**Comment by phadej** [comment]
@ChShersh https://matrix.hackage.haskell.org/package/log-warper looks very sad (*note* different versions fail for different reasons!). `log-warper` on its own seems unusable for `cabal` users. Adding upper-bounds will be very appreciated!

Side comment, `1.8.4` seems to fail to build because it uses `markdown-unlit`. Consider using `build-tool-depends` (http://cabal.readthedocs.io/en/latest/developing-packages.html#pkg-field-build-tool-depends). Looks like `stack` uses that field too: https://github.com/commercialhaskell/stack/blob/a9042ad6fa1d7c813a1c79713a518ee521da9add/src/Stack/Package.hs#L650 (then you don't need to `build-depends` to get `markdown-unlit`, `build-tool-depends` will be enough!). For a "real-world" use see: https://github.com/haskell-servant/servant/blob/0147f4b5c71c70ff136c1d7fea0746cebb34ded7/doc/tutorial/tutorial.cabal#L61 or https://github.com/haskell-servant/servant/blob/0147f4b5c71c70ff136c1d7fea0746cebb34ded7/servant/servant.cabal#L126

**Comment by chshersh** [comment]
@phadej Thanks for reminder regarding `build-tool-depends`! I remember discovering this feature. Will try to use it in `log-warper`.

**Comment by int-index** [comment]
> What do you think of building in Travis with stack and cabal?

Good idea, I added support for `cabal new-build` on Travis in #90

**Comment by chshersh** [comment]
@erikd @phadej @int-index thanks for all your feedback and help! I fixed Travis CI build for `cabal` in PR #91. Cabal CI passes now. I also specified upper bounds for dependencies. And I'm using `build-tool-depends` field for executables. Build matrix on Hackage for `log-warper` now looks a little bit better :slightly_smiling_face: 

* https://matrix.hackage.haskell.org/package/log-warper
